sig-and-wg/sig-core/proposals/ui-api-layer-modularization.md

```diff
@@ -0,0 +1,88 @@
+
+## Goal
+
+1. There should be an ability to turn on and off specific UI API Layer modules, even on server runtime.
+1. If user tries to access queries/mutations/subscriptions of a specific domain, which is related with not installed Kyma component, UI API Layer should return an error about uninstalled domain.
+
+## Suggested solution
+
+The solution is based on a Custom Resource, which is defined for all Kyma components which have equivalents in UI API Layer as different domains (`domain` package). In this proposal it is called as `ConsoleBackendModule`.
+
+Every domain resolver object has to implement `PluggableModule` interface. By default, during start, UI API Layer has all domain resolver disabled, which means there are no informers synced for all resources related to GraphQL queries/mutations/subscriptions.
+
+Using generated Informer API for the `ConsoleBackendModule` resource, an event handler is created for every domain resolver. On create/update/delete events the handler is triggered, and it enables/disables according domain resolver.
+
+Enabling a domain resolver means everything what it is currently done in domain resolver constructor (services, informers setup etc), as well as starting informers and waiting for cache synchronization.
+
+Disabling a specific domain resolver stops the informer synchronization, unregisters subscription listeners and replaces resolver logic to generated one, which always returns an error (e.g. `Module {MODULE_NAME} is not installed`). For all `Container` objects, which contain domain resolvers and exported cross-domain dependencies, all exported fields need to be replaced with generated implementation.
+
+Generator could be based on [Impl](https://github.com/josharian/impl) - but it will need a modification: instead of panicking, it should return error gracefully for every method.
+
+### Architectural Diagram
+
+![UI API Layer Modularization Diagram](assets/ui-api-modularization.svg)
+
+### Custom Resource Definition
+
+The proposed Custom Resource Definition is as follows:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: consolebackedmodules.kyma-project.io
+spec:
+  group: kyma-project.io
+  version: v1alpha1
+  scope: Cluster
+  names:
+    plural: consolebackedmodules
+    singular: consolebackedmodule
+    kind: ConsoleBackendModule
+    shortNames:
+    - cbm
+```
+
+The only required field is unique name of the resource.
+
+Extending existing custom resource, `Microfrontend` and `ClusterMicrofrontend` has been considered. If the `(Cluster)Microfrontend` resource had references to required UI API Layer modules, then we cannot make sure that the Kyma component is actually installed. There could be microfrontends that depends on two components. If just one of them is not installed, it would result in an error/crash.
```

Comment:
```suggestion
Extending the existing custom resources, `Microfrontend` and `ClusterMicrofrontend`, has been considered. If the `(Cluster)Microfrontend` resource had references to the required UI API Layer modules, then we cannot make sure that the Kyma component is actually installed. There could be microfrontends that depend on two components. If just one of them is not installed, it would result in an error/crash.
```